File: server/src/lib/sources/devto.ts
import axios from 'axios'

export interface DevToItem {
  title: string
  url: string
  content: string
  author: string
  publishedAt: Date
  source: 'devto'
  reactions: number
  tags: string[]
  likes?: number
  comments?: number
}

const BASE = 'https://dev.to/api'

// 与 AI/编程领域高度相关的标签
const AI_TAGS = ['ai', 'machinelearning', 'llm', 'deeplearning', 'chatgpt', 'python']

function mapArticle(a: any): DevToItem {
  return {
    title: a.title,
    url: a.url,
    content: a.description || a.title,
    author: a.user?.name || a.user?.username || '',
    publishedAt: new Date(a.published_at),
    source: 'devto' as const,
    reactions: a.public_reactions_count || 0,
    tags: a.tag_list || [],
    likes: a.public_reactions_count || 0,
    comments: a.comments_count || 0,
  }
}

/**
 * 按关键词搜索 Dev.to 文章（用于关键词监控）
 * Dev.to 无公开全文搜索 API，通过相关标签 + 标题匹配实现
 */
export async function searchDevTo(query: string): Promise<DevToItem[]> {
  const queryLower = query.toLowerCase()
  const results: DevToItem[] = []

  // 从与 query 最相关的2个标签中抓取
  const targetTags = AI_TAGS.filter(t => queryLower.includes(t)).slice(0, 2)
  if (targetTags.length === 0) targetTags.push('ai', 'machinelearning')

  for (const tag of targetTags) {
    try {
      const res = await axios.get(`${BASE}/articles`, {
        params: { tag, per_page: 30, state: 'fresh' },
        timeout: 12000,
        headers: { Accept: 'application/json' },
      })

      const articles: any[] = res.data || []
      articles
        .filter(a => {
          const text = `${a.title} ${a.description || ''}`.toLowerCase()
          // 标题或摘要包含查询关键词的首词
          return text.includes(queryLower.split(/\s+/)[0])
        })
        .forEach(a => results.push(mapArticle(a)))
    } catch (err) {
      console.error(`[DevTo] search error for tag=${tag}:`, (err as any)?.message)
    }
  }

  return results
}

/**
 * 获取当天 Dev.to 热门文章（用于热点发现）
 */
export async function getDevToTrending(topic: string): Promise<DevToItem[]> {
  try {
    // top=1 返回最近24小时最受欢迎的文章
    const res = await axios.get(`${BASE}/articles`, {
      params: { per_page: 30, top: 1 },
      timeout: 12000,
      headers: { Accept: 'application/json' },
    })

    const articles: any[] = res.data || []
    const topicLower = topic.toLowerCase().split(/\s+/)[0]

    return articles
      .filter(a => {
        const text = `${a.title} ${a.description || ''} ${(a.tag_list || []).join(' ')}`.toLowerCase()
        // 命中主题关键词，或文章包含 AI 相关标签
        return (
          (topicLower && text.includes(topicLower)) ||
          (a.tag_list || []).some((t: string) => AI_TAGS.includes(t))
        )
      })
      .map(mapArticle)
  } catch (err) {
    console.error('[DevTo] trending error:', (err as any)?.message)
    return []
  }
}

File: .cursor/skills/hot-monitor/scripts/lib/devto.py
from __future__ import annotations

from datetime import datetime, timezone

import requests

BASE = "https://dev.to/api"
AI_TAGS = ["ai", "machinelearning", "llm", "deeplearning", "chatgpt", "python"]
HEADERS = {"Accept": "application/json", "User-Agent": "HotMonitor-Skill/1.0"}


def _map_article(a: dict) -> dict:
    pub_raw = a.get("published_at")
    try:
        published = datetime.fromisoformat(str(pub_raw).replace("Z", "+00:00"))
    except (TypeError, ValueError):
        published = datetime.now(timezone.utc)
    return {
        "title": a.get("title", ""),
        "url": a.get("url", ""),
        "content": a.get("description") or a.get("title", ""),
        "author": (a.get("user") or {}).get("name") or (a.get("user") or {}).get("username") or "",
        "publishedAt": published,
        "source": "devto",
        "likes": a.get("public_reactions_count") or 0,
        "comments": a.get("comments_count") or 0,
    }


def search(query: str) -> list[dict]:
    q_lower = query.lower()
    target_tags = [t for t in AI_TAGS if t in q_lower][:2]
    if not target_tags:
        target_tags = ["ai", "machinelearning"]
    first_word = q_lower.split()[0] if q_lower.split() else q_lower
    results: list[dict] = []
    for tag in target_tags:
        try:
            r = requests.get(
                f"{BASE}/articles",
                params={"tag": tag, "per_page": 30, "state": "fresh"},
                headers=HEADERS,
                timeout=12,
            )
            r.raise_for_status()
            for a in r.json() or []:
                text = f"{a.get('title', '')} {a.get('description', '')}".lower()
                if first_word in text:
                    results.append(_map_article(a))
        except Exception:
            pass
    return results


def trending(topic: str) -> list[dict]:
    topic_lower = (topic.lower().split() or [""])[0]
    try:
        r = requests.get(
            f"{BASE}/articles",
            params={"per_page": 30, "top": 1},
            headers=HEADERS,
            timeout=12,
        )
        r.raise_for_status()
        out: list[dict] = []
        for a in r.json() or []:
            tags = a.get("tag_list") or []
            text = f"{a.get('title', '')} {a.get('description', '')} {' '.join(tags)}".lower()
            if (topic_lower and topic_lower in text) or any(t in AI_TAGS for t in tags):
                out.append(_map_article(a))
        return out
    except Exception:
        return []
